Restore reverted change: Abins refactor replacing SData classes with Euphonic features

ajjackson opened the pull request:
### Description of work

This reinstates #37350 , which was reverted in #39665

This was done in order to stop breakage of package build tests on non-Linux platforms.
It seems this is caused by an incompatibility between multiprocessing "spawn" and the systemtest code, which also spawns Python subprocesses. This mechanism is not supposed to be nested. On Linux the default mechanism is "fork" which seems to work in parallel systemtest, but from Python 3.14 this default will also change to "spawn" and start failing.

#### Summary of work
- The branch from #37350 has been rebased back onto `main`
- The conditions for hanging test suites were investigated
- Multiprocessing code is removed from Abins and replaced by serial `map` operations.
- The performance impact seems to be small, due to optimisations in Euphonic which came after the parallel operations were originally introduced to the branch.

#### Purpose of work
See #37350

#### Further detail of work
It's not great that multiprocessing is incompatible with the systemtest suite. In this case we can manage without, but there is a good chance we will want to use it in future. Multithreading is probably ok, but these are not equivalent tools.

### To test:

Everything is covered by CI. In addition to the default stuff we need to check `build_packages_from_branch` on non-Linux platforms as this was breaking.

This shouldn't require a release note because #37350 didn't.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

ajjackson committed:
Start saving structure factors directly to Spectrum objects

This is a significant step towards removing SData entirely.

For compatibility at this point, Spectrum objects mostly store the bin
centres rather than bin edges. When SData is gone it would nicer to
move to storing bin edges. (Centres can be computed from edges, but
edges can only be guessed from centres...)

ajjackson committed:
Move threshold check, fix 2D systemtest

Bin checks on Spectrum2DCollection need adapting to interpret 3-D
z_data array correctly. This implementation is a bit inefficient as it
needlessly instantiates a Spectrum2D; leave more verbose/efficient
implementation for upstream.

ajjackson committed:
Try to make Spectrum Collection concatenation more efficient

The .from_spectra() constructor has a lot of overhead as it checks the
types and numerical agreement of axes; if we are comparing Collections
this should only be checked once.

If this tweak works well it should move upstream to euphonic. Here we
refer to a private method on the 1D collection from the 2D collection:
really this can live on a parent class.

ajjackson committed:
Apply autoconvolution, q2 corrections to spectra

- create new autoconvolution routine that works with spectrum
  collection

- integrate downsampling into autoconvolution loop over atoms

  - this could save a bit of memory, or benefit from parallism over
    atoms

  - not clear if benefit is large at this point, but it doesn't add
    much complexity either. YAGNI?

- apply q2 order corrections, converting 1D initial spectra to 2D maps
  for quantum orders with isotropic Debye-Waller correction (i.e. all
  orders above fundamental)

- drop _broaden_sdata as this is no longer needed

ajjackson committed:
Tidy up comments and references to SData

ajjackson committed:
Factor out kinematic constraints from SData

This functionality is also available in Euphonic, and it would make
sense to remove this entirely from Abins and use it there.

However
- this will need a bit of tinkering to work with Spectrum2DCollection,
  and doing it upstream would avoid a lot of duplication
- it would be better to apply kinematic constraints (and instrument
  broadening) after spectra have been filtered and summed into output
  groups

So let's defer that for now

ajjackson committed:
Format mass metadata to string

Officially the Spectrum metadata should not include float because they
can't be reliably compared for equality

ajjackson committed:
Tidy up import

ajjackson committed:
Extend broadening efficiency tweak to 2D

Not seeing a big impact on test timings here, but this _should_ scale
better as we add more atoms.

ajjackson committed:
Correct axis-order for dummy 2D spectra

This was breaking a doctest

ajjackson committed:
Return suitable isotropic data quickly, drop empty SData builder

- Rather than go through the data transformation on arrays of zeros,
  we can return an almost-empty SpectrumCollection immediately.
- That finally eliminates SData from this module! There is definitely
  room for optimisation as autoconvolution has become more expensive,
  but the order-1 stuff seems to be running in similar time. It would
  be a good idea to do some profiling later...

- As part of rebasing this, restored some missing units to
  spowdersemiempiricalcalculator and fixed AbinsSDataTest. Sorry, it
  should be on another commit but this is a big job and would benefit
  from some squashing later.

ajjackson committed:
More spowdersemiempiricalcalculator refactoring

- Push SData a bit deeper in the chain of functions

- Stop using SDataByAngle, instead use angle metadata in
  SpectrumCollection objects

- Patch SpectrumCollection so that group_by() and select() work as
  expected (i.e. run succesfully) when some keys are at top level of
  metadata (i.e. common to all lines).

- Apply broadening to SpectrumCollection objects. For now we keep
  using Abins implementation of broadening so that tests pass; this
  should be replaced with Euphonic implementation later.

ajjackson committed:
Continue pushing SData further back into nested function heirarchy

ajjackson committed:
calculate_s_isotropic directly into Spectra

For now, fallback via SData if no isotroptic calculation

ajjackson committed:
Apply Debye-Waller factor to SpectrumCollection, drop SData method

Continue dissolving SData: apply DW to output spectra, drop now-unused
method from class

ajjackson committed:
Extend Euphonic Spectrum types in abins.sdata

This is not very thoroughly tested here, and should be moved upstream
once everything looks happy on Abins side.

- Wrap Spectrum1DCollection to deal with group_by() issue where
  metadata is the same for all items (and so not in line_data)
- Implement Spectrum2DCollection
- Have SData.get_spectrum_collection() dispatch to 1D or 2D as
  appropriate
- Tweak workspace writers to use either SpectrumNCollection class

ajjackson committed:
Refactor threshold check so it can be done outside SData

Move the threshold-check logic to module-level functions, and leave
existing method as a wrapper. Then we should be able to delete the
class later...

Also remove some useless parameters/logic that make the return value
optional. We can just ignore the return value if we don't care about
it.

ajjackson committed:
Change return type of spowdersemiempiricalcalculator

- Now the Algorithm classes never deal with SData at all
- SData still used internally and cached

ajjackson committed:
Drop SData autoconvolution method, port test to replacement function

ajjackson committed:
Implement _atom_type_s using Spectrum1DCollection features

- Although Spectrum1DCollection provides a .select() function, we
  ended up using a chain of filters here because it can't handle
  numerical comparisons. Perhaps this would be a bit cleaner if
  the select() method accepted our filter functions.
- The Spectrum1DCollection group_by() method throws an error if the
  key is in the top level of metadata (i.e. because it applies to all
  rows.) It would be good to clean that upstream too.

ajjackson committed:
Remove Abins SData class

Finally, the cathartic part of the refactor where I get to delete 500
lines of code at once

ajjackson committed:
SData refactor: allow Spectrum1DCollection to be produced

To begin migration from SData to euphonic data types, allow existing
SData to produce a Spectrum1DCollection. Then we can start
modifying code that _uses_ SData to use this datatype instead.

So far we modify the _atom_number_s method which creates output
workspaces from the Abins calculation.

Spectrum2DCollection is not implemented yet, so initially we leave the
old implementation in place for 2D cases.

ajjackson committed:
Rework abins "masses table" to work with spectrum

As the relevant information is available in spectra, this allows us to
drop a redundant parameter and reduce the number of objects passed around

ajjackson committed:
Remove unused SDataByAngle class and tests

The Spectrum1DCollection/Spectrum2DCollection classes allow arbitrary
metadata, so there is no need for a technical distinction between the
class that collects over atoms and quantum orders and one that
collects over angles.

ajjackson committed:
Minor cleanup

ajjackson committed:
More Abins spectra->workspaces tidy-up

ajjackson committed:
Slightly improve performance of spectrum broadening

Remove innermost loop of Spectrum creation by assembling y_data array
directly.

ajjackson committed:
Handle fundamental-only cases more robustly

- The euphonic SpectumCollection select() method doesn't work when all
  lines in the collection match the criterion, because such metadata is
  moved into the top-level metadata rather than the per-line section
  which is checked. The key is therefore treated as missing.

- Compounding this, the data container cannot be entirely
  empty, so it allows an error to be raised if no matches are found.

- More reasonably, the `from_spectra()` constructor cannot operate on
  an empty list, as this means no (common) axis information woudl be
  available. In this commit we avoid calling this when no data is available.

ajjackson committed:
Further optimisation of autoconvolution: Dealing with Pint overhead

Spectrum.x_data (etc.) properties are quite expensive to call as they
trigger a call to ureg to build Units and call Quantity.to() to ensure
output unit is consistent with the stored unit name.

Try to avoid calling it inside a loop! In practice this means using
private attributes a bit more; not ideal, but we pin the Euphonic
version and I maintain the package so for now it is safe.

Avoid calling `* ureg` where we know we have an array and can
build Quantity directly. This saves on some type-checking.

Implement an "unsafe" constructor for Spectrum1DCollection from a
series of Spectrum1D when we already know that the axes will be
consistent. Again this saves a lot of time spent on consistency checks.

ajjackson committed:
Implement iter_metadata for spectrum collections, speed up loops

This iterator makes it more practical to iterate over the content of a
spectrum collection without (slowly) creating new Spectrum objects

Profiling shows a dramatic impact in the places changed so far.

ajjackson committed:
Parallelise Abins autoconvolution over atoms

As the new data container facilitates a more "functional" programming
style with generators, it is quite simple to incorporate
multiprocessing and get the best performance we can from a typical
IDAaaS instance.

Test calculations are showing that with a few cores we get a factor of
2x or so, which seems worth having on what can be a bottleneck
step (if there are a lot of atoms and few k-points, which is typical
for molecular calcualtions.)

I expect that this step will be decoupled from other possible
parallelism over k-points or atom intensity calculations, so it can
use the full available processor count.

ajjackson committed:
More efficient spectrum slicing

Instead of deconstructing to individual spectra, filtering and
reconstructing, just look at the metadata and then index the
collection in one go. This is a lot more efficient as it avoids much
copying of x_data etc.

The [spectrum for spectrum in ...] line was previously taking 70% of
the function run time in a test calculation.
Now it is dominated by _calculate_s_isotropic (97% of time) as it
should be.

ajjackson committed:
Implement parallel broadening over 1D spectra

A simple for-loop over data rows is converted to a Pool.starmap for
an easy speedup.

For a benzene test case on a 30-core EPYC VM, this reduced time in the
function from 1.85s to 0.43s

ajjackson committed:
Tidy up: for now 2D parallel broadening is not implemented

ajjackson committed:
Abins2D optimisation: avoid unnecessary Spectrum iteration

Iteration over metadata is much cheaper than spawning new spectrum
objects.

ajjackson committed:
Experiment with parallel 2D broadening

With my test case of a Benzene cell (one k-point, fair number of
atoms, 10th-order expansion): by default, parallelising the inner loop
over 2D spectra gives a decent speedup as Pool is increased to around
15-16 cores, then deterioriates as we work up to the 30 cores
available.

This seems is significantly improved by setting a chunksize of 5, but we
leave this adjustable in abins.parameters. (There is still a *slight*
edge at some optimal core count, but it is no longer worth the user
effort of checking!)

ajjackson committed:
Better 2D broadening: remove an outer loop

By converting the data to a 2-D array and reshaping at the end, we get
more data into a single starmap() call and better parallel
efficiency. At this point it is also helpful to increase the chunksize
a bit: at 100 the cost increase to 1-D is slight and 2-D benefits a
lot.

Perhaps they should be different parameters and more
benchmarking/adaptive setting used to optimise. But for now this is a
lot better than serial operation!

Performance is now satisfactory, I will not attempt a shared-memory
implementation for the current branch (which is supposed to be focused
on data refactor...)

ajjackson committed:
Tidy up an import statement

ajjackson committed:
Cleaner parallel map with Pool.map and functools.partial

On previous attempt I thought this wouldn't work and that starmap was
required.

The partial function seems a cleaner way of making it clear that only
one parameter is actually being mapped over. It does also seem to work
with a staticmethod, so we can bring the function definition closer to
where it is used.

ajjackson committed:
TMP: try converting parallel map of broadening to series

Getting strange pychop CI errors on Windows and Mac. See if it is
specific to multiprocessing...

ajjackson committed:
Precompute Pychop resolution before parallel map on spectra

Mac/Win tests are failing when the Pychop resolution is computed in
parallel. We don't really need to do that on every thread though; we
can compute the energy function once before distributing.

ajjackson committed:
Create "prepare_resolution()" instrument method

This moves responsibility for method-specific details to the
instruments; at a higher-level we can just call this method to ensure
that we are in a good position for parallel computation.

ajjackson committed:
bump euphonic requirement

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ajjackson committed:
Move type-hint definition to before first use

ajjackson committed:
Revert "bump euphonic requirement"

This reverts commit 75f77f1b214330734697776929e377a264d2af61.

Let's see if the CI will pass before doing this version bump and
cleaning up; currently some code depends on private attributes that
have changed in Euphonic 1.4.

ajjackson committed:
Pin to pre-refactor Euphonic while checking tests on rebase

ajjackson committed:
Bump euphonic requirement

- 1.4.2 fixes an issue with importing base Spectrum

ajjackson committed:
(mostly) delegate Spectrum Collection implementations to Euphonic

I found one bug which we'll patch here for now...

ajjackson committed:
Add missing get_bin_edges method to Spectrum2DCollection

This should be moved upstream but it's helpful to start here and check
on the Mantid CI...

ajjackson committed:
Bump Euphonic, drop patched code

ajjackson committed:
Remove wrappers for Euphonic spectrum classes

ajjackson committed:
Relax Euphonic upper version bound

ajjackson committed:
Use existing euphonic method to iterate over metadata

Spectrum collection store metadata as a mixture of common attributes
and line-specific, e.g.

{"element": "Na",
 "scattering": "Coherent",
 "line_data": [{"atom_index": 1}, {"atom_index": 2}]
}

The constructors and iteration methods on these classes split and
merge as needed. When we want to only iterate over the metadata, it is
wasteful to iterate over the whole collection (i.e. spawning new
Spectrum objects) so we use the iter_metadata() method which doesn't
touch the main data arrays.

This wasn't always available in euphonic so was initially duplicated here.

ajjackson committed:
Quick fixes from code review

- Correct docstring for check_thresholds()

- Avoid unnecessary private access to spectra._z_data when masking
  kinematically-inaccessible values.

  Spectrum objects use a ``@property`` abstraction to store raw arrays
  and units rather than Quantity. This is intended to benefit the
  compiled C extensions by providing a fast private interface to numpy
  arrays that doesn't have to deal with Pint.

  I'm not sure at what point the "set" interface started working
  nicely again, but the classes were recently refactored to use Pint
  more efficiently and a bug may have been fixed in the process.

ajjackson committed:
Rework autoconvolution to use Spectrum1DCollection

This method was working with lists of spectra in order to avoid some
overhead in constructing and iterating over the sequences. Between an
upstream refactor of the .y_data method and the addition of an "usafe"
option to .from_spectra, this should no longer be necessary.

This allows us to work reasonably efficiently without using private
attributes.

ajjackson committed:
Raise error if max autoconvolution order is too low (i.e. = 1)

This is not a likely scenario but would lead to some surprising
behaviour, so let's raise an error.

ajjackson committed:
Review feedback: stricter cast float to string

16 decimal places should be enough to distinguish any Python float

ajjackson committed:
Expand docstring

ajjackson committed:
Review feedback: 1D/2D handling in spowdersemiempiricalcalculator

Clean up these checks by setting a single flag at the level of the
calculator class.

Implemented as a cached property to keep API clean and avoid repeated
lookups from instrument list.

Perhaps the lists could be dispensed with by storing this information
on the instruments themselves. But we would still need a list of them
_somewhere_ for the factory function.

ajjackson committed:
Use already-defined units

ajjackson committed:
Move abins.constants imports to top of file

ajjackson committed:
Simplify abins.parameters.performance["threads"] check

- Use stdlib multiprocessing, which is actually used by the code
- import at top level
- accept None gracefully (the default and recommended setting)

ajjackson committed:
Move another import to top-level

ajjackson committed:
For debugging, tweak package build tests to Abins-only

ajjackson committed:
Force 'spawn' method for multiprocessing - THIS HANGS LINUX SYSTEMTESTS

ajjackson committed:
Quick fix to Abins Systemtest cleanup

Systemtests were not always removing cache files correctly, causing
subsequent runs to skip some of the code.

ajjackson committed:
Remove parallelism from AbINS

Currently multiprocessing interacts poorly with systemtests on
platforms other than linux: we end up with nested multiprocessing and
the tests hang indefinitely.

ajjackson commented:
Aborted systemtest, as expected tweaking the multiprocessing backend caused it to hang on Linux as well.

ajjackson committed:
Tidy up imports

MohamedAlmaki reviewed (changes requested):
Thanks for spotting this tricky bug. The changes remove the multiprocessing code that caused the hanging issue previously. Could you please revert the changes in `run-tests` then do a full `build_packages_from_branch` to make sure it passes in all operating systems before merging into main

ajjackson committed:
Revert "For debugging, tweak package build tests to Abins-only"

This reverts commit 36b203a39323813b66eef59b53607270fa9b5472.

ajjackson commented:
Running here:
https://builds.mantidproject.org/job/build_packages_from_branch/1210/
https://builds.mantidproject.org/job/build_packages_from_branch/1211/
https://builds.mantidproject.org/job/build_packages_from_branch/1212/
https://builds.mantidproject.org/job/build_packages_from_branch/1213/

MohamedAlmaki commented:
Replaced with one build instead of running separate ones 
https://builds.mantidproject.org/job/build_packages_from_branch/1214/

MohamedAlmaki commented:
I have rebuild the packaging because it failed in the first run: https://builds.mantidproject.org/job/build_packages_from_branch/1219/

MohamedAlmaki reviewed (dismissed):
Build passed successfully. Thanks for the fix

ajjackson commented:
It wasn't easy making sense of the log output on the failed run! Looks like 3/4 platforms built and tested fine and the other one barely started?

cailafinn reviewed (changes requested):
Just a couple of code comments on tests and imports:

ajjackson committed:
Move imports to top level

ajjackson committed:
Rework Abins systemtests to use temporary cache directory

For cleaner and safer (parallel) systemtests, run each Abins
systemtest case with its own cache directory and clean up using
tempfile mechanisms.

cailafinn reviewed (approved):
This looks good to me now, thanks!
